<a href="https://colab.research.google.com/github/KartheekAnusuri/Applied-Tech/blob/main/CapstoneProject7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project 7: Bengaluru House Prices - Data Cleaning


---

### Context

A techie residing in Mumbai who is a data scientist got an excellent job opportunity from one of his dream company in Bengaluru. Now, he wants to shift to Bengaluru soon and decides to buy a home in the city.
The lack of trust in property developers in the city resulted in a drop of 7% in the housing units sales across India in 2017. The property prices in Bengaluru fell by almost 5% in the second half of 2017, said a study published by a property consultancy firm called Knight Frank. Here's the link to the news article:

[Residential real estate prices drop 3% in big cities
](https://www.businesstoday.in/sectors/infra/residential-real-estate-prices-property-drop-3-per-cent-realty/story/267752.html)

Buying a home, especially in a city like Bengaluru, is a tricky choice. While the major factors are usually the same for all metro cities, there are others to be considered for the Silicon Valley of India. Because of the millennial crowd, vibrant culture, great climate and a slew of job opportunities in Bengaluru, it is difficult to ascertain the price of a house in the city.
As he is a data scientist he finds the dataset for Bengaluru house prices on the internet.

The dataset he found was vast with some useless information and some of the empty values, he only wants to extract some of the useful information from the dataset.

---

### Problem Statement

The dataset acquired by the techie is full of irregularities, incorrect values and missing values. As a data scientist (or analyst in this context), your task is to clean the dataset given to you in this problem statement.

This process of preparing data for analysis by removing or modifying data that is incorrect, incomplete, irrelevant, duplicated, or improperly formatted is known as data cleaning.


---

---

### Dataset Information

This dataset contains several factors that a potential home buyer considers before purchasing a house. They are the following:

1. `area_type`: Whether the house has a built-up area or plot area.

2. `availability`: Whether the house is in the ready-to-move state or the exact date on which the house will be ready for occupancy.

3. `location`: The name of the locality in Bengaluru where the house is situated.

4. `size`: Whether the house is 1 BHK, 2 BHK, 3 BHK etc. The term BHK stands for *Bedroom, Hall & Kitchen*.

5. `society`: Name of the society where the house is located.

6. `total_sqft`: The total area of the house in square feet.

7. `bath`: Number of bathrooms in the house.

8. `balcony`: Number of balconies in the house.

9. `price`: Price of the house in lakhs (Indian rupees).


---

### Things To Do

- Create a Pandas DataFrame for the Bengaluru house prices dataset.

- Find the missing values in the DataFrame and replace them with the right missing values. However, if a column has more than 15% missing values then drop (or remove) the column from the DataFrame except for the `location, size` and `total_sqft` columns because the house prices are directly dependent on these three crucial parameters. *In fact, the rate of a house is reported as a cost per unit area of the house*.

- Drop (or remove) the rows containing the missing values in the `location,	size` and `total_sqft` columns. Additionally, drop any row which contains nonsensical values in the context of houses.

---

#### 1. Import Modules

First import all the necessary modules to create a DataFrame and box plots.

In [ ]:
# Import the necessary modules.
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

---

#### 2. Load The Dataset

Create a Pandas DataFrame by loading the dataset which is provided in the comma-separated values (CSV) file. Also, display the number of rows and columns in the DataFrame. Here's the link to the dataset:

https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/bengaluru-house-prices/Bengaluru_House_Prices.csv

In [ ]:
# Load the dataset.
df = pd.read_csv("https://student-datasets-bucket.s3.ap-south-1.amazonaws.com/whitehat-ds-datasets/bengaluru-house-prices/Bengaluru_House_Prices.csv")

In [ ]:
# Display the number of rows and columns.
num_rows, num_cols = df.shape
print("Number of rows:", num_rows)
print("Number of columns:", num_cols)

Number of rows: 13320
Number of columns: 9


---

#### 3. Check For The Missing Values

- Compute the total number of missing values in the DataFrame.

- Compute the percentage of missing values in the DataFrame. If a column has more than 15% missing values then drop (or remove) the column from the DataFrame.

In [ ]:
# Find the total number of missing values in each column.
df.isnull().sum()

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64

**Hint**: You can find the total number of missing values in each column by using the `sum()` function on top of the `isnull()` function.

**Q:** Which column has the most number of missing values?

**A:** Society

In [ ]:
# Find the percentage of missing values
df.isnull().mean() * 100

area_type        0.000000
availability     0.000000
location         0.007508
size             0.120120
society         41.306306
total_sqft       0.000000
bath             0.548048
balcony          4.572072
price            0.000000
dtype: float64

**Hint** : You can calculate the percentage of missing values in each column by multiplying the total number of missing values in each column with 100 and dividing the value with the total number of rows in the DataFrame.

**Q:** Which is/are column(s) that need to be removed? Write "Not Applicable" if there are none.

**A:** Society and balcony

---

### 4. Data Cleaning - Rows

First things first, drop (or remove) the rows containing the missing values in the `location,	size` and `total_sqft` columns because they are the most important parameters in governing the price of a house.

---

#### 4.1 Removing Rows From The `location` Column

You need to remove all the rows containing the missing values in the `location` column.

In [ ]:
# Get all the rows having the missing values in the 'location'.
df.dropna(subset=['location'], inplace=True)

**Q:** How many rows have the missing values in the `location` column?

**A:** 0

In [ ]:
# Total number of rows having the missing values in the 'location' column.
df['location'].isnull().sum()

0

In [ ]:
# Discard the rows containing the missing values in the 'location' column.
df.dropna(subset=['location'], inplace=True)

**Hint** : Use `isnull()` function inside the `loc[]` function to locate and remove the rows with missing values from the `location` column.

---

#### 4.2 Removing Rows From The `size` Column

Now, you need to remove all the rows containing the missing values in the `size` column.

In [ ]:
# Get the rows having the missing values in the 'size' column.
df.dropna(subset=['size'], inplace=True)

**Q:** How many rows have the missing values in the `size` column?

**A:** 0

In [ ]:
# Total number of rows having the missing values in the 'size' column.
df['size'].isnull().sum()

0

In [ ]:
# Discard the rows containing the missing values in the 'size' column.
df.dropna(subset=['size'], inplace=True)

**Q:** How many rows are there in the DataFrame after removing the rows that need to be removed?

**A:**

---

#### 4.3 Removing Rows From The `total_sqft` Column

Now, you need to remove all the rows containing the missing values in the `total_sqft` column.

In [ ]:
# Get the rows having the missing values in the 'total_sqft' column.
df.dropna(subset=['total_sqft'], inplace=True)

**Q:** How many rows have the missing values in the `size` column?

**A:** 0

---

#### 4.3 Removing Rows From The `bath` Column

Assume that in the practical situation, no house, even in a city like Bengaluru, will have more than 5 bathrooms. Hence, drop all the rows in the `bath` column containing more than 5 bathrooms.

In [ ]:
# Get the rows having more than 5 bathrooms in the 'bath' column.
df.dropna(subset=['bath'], inplace=True)

In [ ]:
df[df['bath'] > 5].shape[0]

515

**Q:** How many rows have more than 5 bathrooms in the `bath` column?

**A:** 515

In [ ]:
# Discard the rows having more than 5 bathrooms in the 'bath' column.
df[df['bath'] <= 5]

area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath  balcony   price  
0          2 BHK  Coomee        1056   2.0      1.0   39.07  
1      4 Bedroom  Theanmp       2600   5.0      3.0  120.00  
2          3 BHK      NaN       1440   2.0      3.0   62.00  
3          3 BHK  Soiewre       1521   3.0      1.0   95.00  
4          2 BHK      NaN       1200   2.0      1.0   51.00  
...          ...      ...        ...   ...      ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0      0.0  231.00  
13316      4 BHK      NaN       3600   5.0      NaN  400.00  
13317      2 BHK  Mahla T       1141   2.0      1.0   60.00  
13318      4 BHK  SollyCl       4689   4.0      1.0  488.00  
13319      1 BHK      NaN        550   1.0      1.0   17.00  

[12731 rows x 9 columns]

**Q:** How many rows have more at most 5 bathrooms in the `bath` column?

**A:** 515

---

### 5. Data Cleaning - Columns

Compute the percentage of missing values in the DataFrame again. If a column has more than 15% missing values then drop (or remove) the column from the DataFrame.


In [ ]:
# Percentage of missing values.
missing_percent = (df.isnull().mean() * 100).round(2)
missing_percent

area_type        0.00
availability     0.00
location         0.00
size             0.00
society         41.51
total_sqft       0.00
bath             0.00
balcony          4.05
price            0.00
dtype: float64

**Hint**: You can calculate the percentage of missing values in each column by multiplying the number of missing values with 100 and dividing the resulting value with total number of rows in the DataFrame.

**Q:** Which is/are column(s) that need to be removed? Write "Not Applicable" if there are none.

**A:** society and balcony need to be removed.

**Q:** Did the overall percentage of the missing values dropped compared to the initial percentage of the missing values? Write yes or no.

**A:** yes

---

#### 5.1 Columns To Be Retained

Create a list of columns that need to be retained.

In [ ]:
# List of the columns to be retained.
columns_to_retain = []
for column in df.columns:
    if missing_percent[column] <= 15:
        columns_to_retain.append(column)

print("Columns to be retained:")
print(columns_to_retain)

Columns to be retained:
['area_type', 'availability', 'location', 'size', 'total_sqft', 'bath', 'balcony', 'price']


**Hint**: You need to create a list of columns that have missing values not more than 15%. This can be done by appending such column names in a list using a `for` loop.

---

#### 5.2 Retaining Columns

Retain the appropriate columns in the DataFrame.

In [ ]:
# Retain the appropriate columns in the DataFrame.
df_retained = df.loc[:, columns_to_retain]

**Hint**: You need to create a DataFrame with columns having missing values not more than 15%. This can be done by passing the variable holding the list of columns inside a `loc[]` function.

Again calculate the percetange of missing values in the DataFrame.

In [ ]:
# Percentage of the missing values in the DataFrame.
missing_percent_after = (df_retained.isnull().mean() * 100).round(2)
missing_percent_after

area_type       0.00
availability    0.00
location        0.00
size            0.00
total_sqft      0.00
bath            0.00
balcony         4.05
price           0.00
dtype: float64

**Q:** Write down (or identify) the column(s) in which the values need to be replaced with an appropriate value.

**A:** missing_percent_after[missing_percent_after > 0].index retrieves the columns from the updated DataFrame df_retained that still have missing values. These columns are identified by checking if their percentage of missing values (missing_percent_after) is greater than 0.

---

---